app/sprinkles/account/src/Controller/AccountController.php

```diff
@@ -405,6 +429,38 @@ public function login(Request $request, Response $response, $args)
 
         $currentUser = $authenticator->attempt(($isEmail ? 'email' : 'user_name'), $userIdentifier, $data['password'], $data['rememberme']);
 
+        $passwordSecurity = $this->ci->passwordSecurity;
+
+        // Check if the enforce password update setting is configured.
+        if ($passwordSecurity->resetCompromisedEnabled()) {
+            // Check if the password is on the compromised password list.
+            $numberOfBreaches = $passwordSecurity->checkPassword($data['password']);
+
+            if ($passwordSecurity->breachThreshold() != '-1' && $numberOfBreaches > $passwordSecurity->breachThreshold()) {
+
+                // Try to generate a new password reset request.
+                // Use timeout for "reset password"
+                $passwordReset = $this->ci->repoPasswordReset->create($currentUser, $config['password_reset.timeouts.reset']);
+                $token = ['token'=> $passwordReset->getToken()];
+
+                // Load validation rules - note this uses the same schema as "set password"
+                $schema = new RequestSchema('schema://requests/set-password.yaml');
+                $validator = new JqueryValidationAdapter($schema, $this->ci->translator);
+
+                $forcePasswordChange = $this->ci->router->pathFor('set-password', [
+                'page' => [
+                  'validators' => [
+                      'set_password'    => $validator->rules('json', false)
+                  ]
+                ]
+              ], $token);
+
+                $ms->addMessageTranslated('info', 'PASSWORD.SECURITY.RESET_REQUIRED.COMPROMISED');
+
+                return $response->withRedirect($forcePasswordChange);
```

Comment:
@Silic0nS0ldier I'm going to expand on this here since the code changed a lot since my first comment. 

**Expected behavior:** when `site.login.enforce_reset_compromised` is set to `true` then the user's password is checked at each log in. If the password is found to be unacceptable according to the config values, the user is expected to be redirected to the password reset page `return $response->withRedirect($forcePasswordChange);` and the `login` function should exit.

**Actual behavior:** The controller is returning both of the return statements at L460 and L469. You can see this in the Fiddler screenshot and based off that fact that the message returned here at L460 ends up displaying on the homepage in the screen shot below. 
```
return $response->withRedirect($forcePasswordChange);

        return $redirectOnLogin($request, $response, $args);
```


![image](https://user-images.githubusercontent.com/33728190/58181067-1cf1b380-7c79-11e9-9a93-6618ccc40897.png)

![image](https://user-images.githubusercontent.com/33728190/58181101-2b3fcf80-7c79-11e9-9945-ce1861a62757.png)


To reproduce, you should be able to checkout this branch and try to log in a user with a "bad" password, such as `password1234`

Reply:
Thanks. I've managed to reproduce the issue so I'll into this further once I have the chance. Depending on how tricky this proves to be, I should have an answer before the end of my work break.

Reply:
Looks to be some issues in client side scripts causing this.

From what I can see, what should be happening is that when someone logs in with an exposed credential they are redirected to the set password page? If yes then that raises the question of what is the best way to handle this. If we have evidence of exposed credentials would be not want to require the user to verify their identity via email and outright reject login?

Reply:
> From what I can see, what should be happening is that when someone logs in with an exposed credential they are redirected to the set password page?

Yes this is how I intended for it to work. 

> If yes then that raises the question of what is the best way to handle this. If we have evidence of exposed credentials would be not want to require the user to verify their identity via email and outright reject login?

Well, maybe. There are at least two situations to consider.
1. This feature is activated on an existing UserFrosting site and as a result a large number of users may be required to update their passwords. 

2. The feature is already active on a site and a new breach is added to Have I Been Pwned list, flagging a user's password as now being unacceptable for use. 

In either situation it does not necessarily mean that the UserFrosting site's username/passwords have been exposed, just that in some other breach the password has been exposed.

I do understand your point about verifying their identity though. Forcing the user to verify their identity would be the more secure option, if only slightly more secure. As always it comes down to whether user experience or security is more important.

 If we approach this as a standard password rule, would it be normal to verify a user before allowing them to update their password if say, the minimum password length was raised from 8 characters to 10?

Reply:
I think a situation where there is a possibility that someones password has been exposed we would want to explicitly **prevent** a quick and easy reset. Otherwise we would just be helping a would be attacker to lock a user out of their account. The danger goes up a lot of the account hijacked is that of the admin.

Other scenarios we could implement later like password expiration or changes to complexity requirements would be part of a different discussion.

Reply:
What you are suggesting will diminish user experience and I am not convinced it will actually result in better security. Right now, users can reset their passwords via the account settings page. If there is benefit from forcing a user to verify their identity prior to resetting their password we should not be allowing password resets in the application at all - only via password reset email. 

Consider that there is a data breach of some company and an email and password are exposed by that breach that are also being used in a UF application. By the time said data breach actually makes its way to being listed on HIBP site (After being announced by breached company or attackers, and then undergoing HIBP verification, vetting, etc..) the attacker has already had time to log into the UF application and reset the user's password via the User Settings page.


> Other scenarios we could implement later like password expiration or changes to complexity requirements would be part of a different discussion.

I am not trying to discuss implementing those. I consider the password blacklist to fall into the same category as those password constraints/requirements, thus asking the question: 
> If we approach this as a standard password rule, would it be normal to verify a user before allowing them to update their password if say, the minimum password length was raised from 8 characters to 10?

I think we could use some outside input on this. I am happy to head in the proposed direction if there is consensus it is the best option. 
@lcharette @alexweissman

Reply:
According to [this article](https://www.owasp.org/index.php/Testing_for_weak_password_change_or_reset_functionalities_(OTG-AUTHN-009)), and IF this feature falls under the category of a `password change` rather than a `password reset`, the question to ask is:

> - Is the old password requested to complete the change?

> The most insecure scenario here is if the application permits the change of the password without requesting the current password. Indeed if an attacker is able to take control of a valid session they could easily change the victim's password.

Though 'm not entirely sure whether redirecting the user to the page after they login covers the requirements of "requiring current password to allow password change".